islandora_compound_object.module

```diff
@@ -268,17 +268,94 @@ function islandora_compound_object_access(AbstractObject $object) {
 }
 
 /**
- * Replaces parent object thumbnail with child's thumbnail.
+ * Implements hook_compound_cmodel_islandora_derivative().
  */
-function islandora_compound_object_update_parent_thumbnail($parent) {
+function islandora_compound_object_compound_cmodel_islandora_derivative(AbstractObject $object = NULL, $ds_modified_params = array()) {
+  $module_path = drupal_get_path('module', 'islandora_solution_pack_compound');
+  return array(
+    array(
+      'source_dsid' => NULL,
+      'destination_dsid' => 'TN',
+      'function' => array('islandora_compound_object_update_parent_thumbnail'),
+      'file' => "$module_path/islandora_compound_object.module",
+    ),
+  );
+
+}
+
+/**
+ * Implements hook_islandora_derivative().
+ */
+function islandora_compound_object_islandora_derivative(AbstractObject $object) {
+  return array(
+    array(
+      'source_dsid' => 'TN',
+      'destination_dsid' => NULL,
+      'function' => array('islandora_compound_object_update_parent_thumbnail_from_child'),
+    ),
+  );
+}
+
+/**
+ * Update parent from child object.
+ *
+ * Given a child object, check to see if it is part of
+ * a compound and update the parent's TN datastream.
+ *
+ * @param AbstractObject $child
+ *   The potential child object.
+ */
+function islandora_compound_object_update_parent_thumbnail_from_child(AbstractObject $child) {
   module_load_include('inc', 'islandora', 'includes/utilities');
+  // Short circuit if feature is disabled.
+  if (!variable_get('islandora_compound_object_tn_deriv_hooks', FALSE)) {
+    return;
+  }
+  // Test if child has a parent.
+  $compound_info = islandora_compound_object_retrieve_compound_info($child);
+  if (empty($compound_info)) {
+    return;
+  }
+  // Test if child is the first child of the compound.
+  if (empty($compound_info['previous_pid'])) {
+    $parent = islandora_object_load($compound_info['parent_pid']);
+    // Child doesn't have a thumbnail then we can't set the parents TN either.
+    if (empty($child['TN'])) {
+      // Parents with first children with no TN shouldn't have one either.
+      if (isset($parent['TN'])) {
+        $parent->purgeDatastream('TN');
+      }
+      return;
+    }
+    islandora_compound_object_update_parent_thumbnail($parent, $child);
+  }
+}
+
+/**
+ * Update a parent object's TN by deriving it's first child's TN.
+ *
+ * @param AbstractObject $parent
+ *   The compound object.
+ * @param AbstractObject $child
+ *   The child object (optional).
+ */
+function islandora_compound_object_update_parent_thumbnail(AbstractObject $parent, AbstractObject $child = NULL) {
+  // Short circuit if feature is disabled.
+  if (!variable_get('islandora_compound_object_tn_deriv_hooks', FALSE)) {
```

Comment:
Short-circuiting stuff can now be stripped from here, since [it's now being handled elsewhere](https://github.com/Islandora/islandora_solution_pack_compound/pull/126/files#diff-dacfa496dcdc86ad0174e734682f128bR296).